FIX: install/update app with permissions on b2g instance

rpl opened the pull request:
This change hooks the new remote webappsActor to install apps into a b2g instance
using mozApps internals, so application permissions will be registered and
correctly available.

This merge should fix: #237

rpl committed:
FIX: use new remote webappsActor to install app with permissions on b2g instance

rpl committed:
FIX: fix installing hosted and hosted generated apps using remote webappsActor

digitarald commented:
Hooray!

Only issue I run into when using the actor (from my own xpcshell script), was that re-installs didn't work as expected. Like adding/removing fullscreen or other properties from the manifest wasn't recognized until I uninstalled the app and rebooted the phone.

rpl commented:
Currently I left intact the method we use to remove applications and I get a reasonable doubt that in this way some information (such as permissions that are registered and de-registered during the installation and uninstallation) will be left in the system.

I need to go a bit deeper (although in the end the real solution is surely to provide the opposite function into an debugger server actor)

mykmelez commented on addon/lib/remote-simulator-client.js:
```diff
@@ -225,6 +226,15 @@ const RemoteSimulatorClient = Class({
+
+    remote.client.request({to: remote.webapps, type: "install", appId: appId,
+                           appType: appType}, 
+                          onResponse);
```
Nit: this would be easier to read if the object argument was more structured:

```
remote.client.request({ to: remote.webapps,
                        type: "install",
                        appId: appId,
                        appType: appType },
                      onResponse);
```

That also makes the lines short enough to enable you to call request() directly on this._remote.client without exceeding 80 characters per line, so you don't have to assign it a local variable that then only gets used once:

```
this._remote.client.request({ to: remote.webapps,
                              type: "install",
                              appId: appId,
                              appType: appType },
                            onResponse);
```

mykmelez commented on addon/lib/main.js:
```diff
@@ -126,10 +126,19 @@ let simulator = {
+    XPCOMUtils.defineLazyServiceGetter(sandbox, "gDirService",
+                                       "@mozilla.org/file/directory_service;1",
+                                       "nsIProperties");
+    return File.join(URL.toFilename("file://"+sandbox.gDirService.get("TmpD", Ci.nsILocalFile).path), "b2g");
```
I don't understand this line of code. `File.join()` takes two path parts, and this generates the first by 1. getting a path, 2. converting it to a URL via string concatenation, 3. converting it back to a path via `URL.toFilename()`. But isn't the first step sufficient? I.e. simply:

```
return File.join(sandbox.gDirService.get("TmpD", Ci.nsILocalFile).path, "b2g");
```

If you really do need to convert the path to a URL, you should use `URL.fromFilename()` to do it, which would make this:

```
return File.join(URL.toFilename(URL.fromFilename(sandbox.gDirService.get("TmpD", Ci.nsILocalFile).path)), "b2g");
```

But then you should really add a code comment explaining why it's necessary to convert the path to a URL only to turn right around and convert it back again. And wrap the line so it doesn't exceed 80 characters in length.

mykmelez replied:
One more nit: nsILocalFile is deprecated, so use nsIFile instead.

mykmelez replied:
Erm, and I just realized: _main.js_ imports _Services.jsm_, which provides _dirsvc_, so we don't need to get it ourselves, and this can be as simple as:

```
get tempDir() {
  return File.join(Services.dirsvc.get("TmpD", Ci.nsIFile).path, "b2g");
},
```

mykmelez commented on addon/lib/main.js:
```diff
@@ -126,10 +126,19 @@ let simulator = {
     }
   },
 
+  get webappsDir() {
```
Nit: it would be clearer to call this getter something like _tempDir_, since there are other directories in this codebase that are called _webappsDir_.

mykmelez commented on addon/lib/main.js:
```diff
@@ -173,90 +182,100 @@ let simulator = {
+        
+          // Complete install (Hosted)
+          if (manual) {
+            simulator.defaultApp = null; // id; // DISABLED: because on the first run the app will be not already installed
```
Nit: same as for above comment.

mykmelez commented on addon/lib/main.js:
```diff
@@ -664,12 +691,7 @@ let simulator = {
-          this.remoteSimulator.once("ready", function() cmd());
-          this.run();
-        }
+        this.run(cmd);
```
Nice simplification! It makes more sense to encapsulate this logic in the `run()` function.

rpl committed:
FIX: workaround appId generation regression (due to b2g-desktop forced exit) and code cleanup

rpl committed:
change gDirService to Services.dirsvc

rpl committed:
don't kill b2g instance to install and update apps

rpl commented:
I've introduced suggested code cleanup and fixed a regression (due to new app install method and b2g install forced kill)

In the last commit I've removed from tab worker message handler any simulator.kill used from the previous install/upgrade method.

I've not been yet able to study more deeply potential problems related to the cases mentioned by @digitarald

mykmelez commented:
Thanks for the fixes!
